src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,171 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private int size = 0;
+    private Node<T> head;
+    private Node<T> tail;
+
+    private static class Node<T> {
+        T value;
+        Node<T> next;
+        Node<T> prev;
+
+        public Node(Node<T> prev, T value, Node<T> next) {
+            this.prev = prev;
+            this.value = value;
+            this.next = next;
+        }
+    }
+
     @Override
     public boolean add(T value) {
-        return false;
+        addToLast(value);
+        return true;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == size) {
+            addToLast(value);
+        } else {
+            addBefore(value, getCurrentNode(index));
+        }
+    }
+
+    private void addToLast(T value) {
+        Node<T> lastNode = tail;
+        Node<T> newNode = new Node<>(lastNode, value, null);
+        tail = newNode;
+        if (lastNode == null) {
+            head = newNode;
+        } else {
+            lastNode.next = newNode;
+        }
+        size++;
+    }
+
+    private void addBefore(T value, Node<T> currentNode) {
+        Node<T> beforeNode = currentNode.prev;
+        Node<T> newNode = new Node<>(beforeNode, value, currentNode);
+        currentNode.prev = newNode;
+        if (beforeNode == null) {
+            head = newNode;
+        } else {
+            beforeNode.next = newNode;
+        }
+        size++;
+    }
 
+    public Node<T> getCurrentNode(int index) {
+        rangeCheck(index);
+
+        Node<T> currentNode = head;
+        for (int i = 0; i < index; i++) {
+            currentNode = currentNode.next;
+        }
+        return currentNode;
+    }
+
+    private void rangeCheck(int index) {
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException("Index more than size or less than 0");
+        }
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        for (int i = 0; i < list.size(); i++) {
+            T value = list.get(i);
+            add(value);
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        rangeCheck(index);
+        if (index == size) {
+            throw new IndexOutOfBoundsException("Index equal size");
+        }
+        return getCurrentNode(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (index == size) {
+            throw new IndexOutOfBoundsException("Index equal size");
+        }
+        rangeCheck(index);
```

Comment:
inside of `getCurrentNode(index)` method you already have `rangeCheck(index);` calls. Can we remove this redundant call?

Reply:
Yes